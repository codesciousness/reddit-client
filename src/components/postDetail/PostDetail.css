.PostDetail {
    width: 80%;
    margin: auto;
}