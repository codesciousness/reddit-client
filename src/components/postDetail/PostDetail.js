import React from 'react';
import { useSelector } from 'react-redux';
import './PostDetail.css';
import Post from '../post/Post';
import Comments from '../../features/comments/Comments';
import { selectTrendingPosts, selectLatestPosts, selectCurrentPost } from '../../features/posts/postsSlice';

const PostDetail = () => {
    const currentPost = useSelector(selectCurrentPost);
    const trendingPosts = useSelector(selectTrendingPosts);
    const latestPosts = useSelector(selectLatestPosts);

    let post;
    trendingPosts.forEach(postObj => {
        if (postObj.id === currentPost) {
            post = postObj;
        }
    });
    latestPosts.forEach(postObj => {
        if (postObj.id === currentPost) {
            post = postObj;
        }
    });

    if (post) {
        return (
            <section className="PostDetail">
                <Post post={post} isTrending={false} />
                <Comments />
            </section>
        );
    }
    else return null;
}

export default PostDetail;