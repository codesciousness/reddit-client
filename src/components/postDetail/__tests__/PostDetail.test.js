import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PostDetail from '../PostDetail';

jest.mock('../../../features/comments/Comments', () => () => {
    return (
        <div data-testid="Comments">
            <p>Comments</p>
            <p>Very informative article! | Posted By: ScienceReader</p>
            <p>Future research on this topic should definitely move more in this direction. | Posted By: ScienceLearner</p>
        </div>
    );
});

const mockStore = configureStore([]);

const trendingPosts = [
    {
        id: 'efgh2i',
        title: 'Where Can the Solar Eclipse be Viewed Next Week',
        author: 'ScienceResearcher',
        created: '4 hours ago',
        flair: 'Astronomy',
        thumbnail: 'https://source.unsplash.com/140x140/?astronomy',
        commentNum: 25,
        url: 'https://www.sciencemag.org/'
    },
    {
        id: 'jklm3n',
        title: 'An Overview of Traditional Diets versus Modern Diets',
        author: 'ScienceLover',
        created: '2 hours ago',
        flair: 'Health',
        thumbnail: 'https://source.unsplash.com/140x140/?health',
        commentNum: 6,
        url: 'https://www.nutrition.org/'
    }
];

const latestPosts = [
    {
        id: 'opqr4s',
        title: 'Causes of Reproductive Stress on Animals in Declining Habitats',
        author: 'ScienceGuy',
        created: '25 minutes ago',
        flair: 'Biology',
        thumbnail: 'https://source.unsplash.com/140x140/?biology',
        commentNum: 2,
        url: 'https://www.nature.com/'
    },
    {
        id: 'tuvw5x',
        title: 'Activation of the Brain in Subjects With High And Average Intelligence',
        author: 'ScienceObserver',
        created: '40 minutes ago',
        flair: 'Neuroscience',
        thumbnail: 'https://source.unsplash.com/140x140/?neuroscience',
        commentNum: 3,
        url: 'https://www.jneurosci.org/'
    }
];

describe('<PostDetail />', () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            posts: {
                currentPost: 'jklm3n',
                trendingPosts,
                latestPosts
            }
        });
    });

    test('PostDetail component is defined', () => {
        expect(<PostDetail />).toBeDefined();
    });
    test('matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <PostDetail />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly when currentPost is truthy', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <PostDetail />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.PostDetail')).toBeInTheDocument();
    });
    test('returns null when currentPost is falsy', () => {
        store = mockStore({
            posts: {
                currentPost: '',
                trendingPosts,
                latestPosts
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <PostDetail />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.PostDetail')).toBeNull();
    });
    test('renders a Post component with isTrending=false prop passed', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <PostDetail />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Post__latest')).toBeInTheDocument();
        expect(screen.getByText('An Overview of Traditional Diets versus Modern Diets')).toBeInTheDocument();
        expect(screen.getByText('ScienceLover')).toBeInTheDocument();
        expect(screen.getByText('Health')).toBeInTheDocument();
    });
    test('renders a Comments component', () => {
        render(
            <Provider store={store}>
                <Router>
                    <PostDetail />
                </Router>
            </Provider>
        );
        expect(screen.getByTestId('Comments')).toBeInTheDocument();
        expect(screen.getByText(/Very informative article/)).toBeInTheDocument();
        expect(screen.getByText(/ScienceLearner/)).toBeInTheDocument();
    });
});