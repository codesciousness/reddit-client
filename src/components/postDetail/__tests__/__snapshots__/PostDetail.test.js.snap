// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostDetail /> matches the snapshot 1`] = `
<section
  className="PostDetail"
>
  <section
    className="Post__latest"
  >
    <img
      alt=""
      className="Post__latest__image"
      src="https://source.unsplash.com/140x140/?health&sig=jklm3n"
    />
    <div
      className="Post__latest__content"
    >
      <p
        className="Post__latest__flair"
      >
        Health
      </p>
      <a
        className="Post__latest__link"
        href="/posts/health/jklm3n"
        onClick={[Function]}
      >
        <h2
          className="Post__latest__title"
          id="jklm3n"
          onClick={[Function]}
        >
          An Overview of Traditional Diets versus Modern Diets
        </h2>
      </a>
      <a
        href="https://www.nutrition.org/"
        rel="noreferrer"
        target="_blank"
      >
        <button
          className="Post__latest__button"
        >
          Read more 
          <i
            className="fas fa-angle-double-right Post__latest__icon"
          />
        </button>
      </a>
      <p
        className="Post__latest__detail"
      >
        <span
          className="Post__latest__author"
        >
          Posted by 
          <strong>
            ScienceLover
          </strong>
        </span>
        <span
          className="PostTime"
        >
          NaN years ago
        </span>
        <span
          className="Post__latest__comments"
        >
          <i
            className="far fa-comment-alt"
          />
           
          6
        </span>
      </p>
    </div>
  </section>
  <div
    data-testid="Comments"
  >
    <p>
      Comments
    </p>
    <p>
      Very informative article! | Posted By: ScienceReader
    </p>
    <p>
      Future research on this topic should definitely move more in this direction. | Posted By: ScienceLearner
    </p>
  </div>
</section>
`;
